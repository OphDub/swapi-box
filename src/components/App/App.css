.app {
  color: white;
  background-color: black;
  display: flex;
  flex-direction: column;
  align-items: center; }

header {
  display: flex;
  flex-direction: row;
  justify-content: center;
  width: 100%; }

.swapi-title {
  display: flex;
  align-self: center; }
