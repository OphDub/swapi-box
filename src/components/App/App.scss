html {
  height: 100%;
  width: 100%;
  margin: 0px;
  padding: 0px;
}

body {
  height: 100%;
  width: 100%;
  margin: 0px;
  padding: 0px;
}

#root {
  min-height: 100%;
  background-color: magenta;
  display: flex;
}

.app {
  color: white;
  background-color: black;
  display: flex;
  flex-direction: column;
  align-items: center;
  margin: 0px;
  padding: 0px;
  flex: 1;
}

header {
  display: flex;
  flex-direction: row;
  justify-content: center;
  width: 100%;
}

.swapi-title {
  display: flex;
  align-self: center;
}
