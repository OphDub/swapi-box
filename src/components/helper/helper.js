export const cleanFilmData = (filmData) => {
  return filmData.results.map(film => {
    return {
      title: film.title,
      crawl: film.opening_crawl,
      releaseDate: film.release_date,
    }
  })
}

const getHomeworldData = async (url) => {
  const initialFetch = await fetch(url)
  const homeworldObj = await initialFetch.json()

  return {
    homeworld: homeworldObj.name,
    population: homeworldObj.population,
  }
}

const getSpeciesData = (urls) => {
  const unresolvedPromises = urls.map(async (url) => {
    const initialFetch = await fetch(url)
    const species = await initialFetch.json()

    return species.name
  })
  return Promise.all(unresolvedPromises);
}

export const cleanPeopleData = (peopleData) => {
  const people = peopleData.results.map( async (person) => {
    const homeworld = await getHomeworldData(person.homeworld)
    const speciesTypes = await getSpeciesData(person.species)

    return {
      name: person.name,
      ...homeworld,
      species: speciesTypes,
    }
  })
  return Promise.all(people);
}

export const cleanVehicleData = (vehicleData) => {
  return vehicleData.results.map((vehicle) => {
    return {
      name: vehicle.name,
      model: vehicle.model,
      passengers: vehicle.passengers,
      vehicleClass: vehicle.vehicle_class,
    }
  })
}

export const getPlanetResidents = (urls) => {
  const unresolvedPromises = urls.map( async (url) => {
    const initialFetch = await fetch(url)
    const resident = await initialFetch.json()

    return resident.name
  })

  return Promise.all(unresolvedPromises)
}

export const cleanPlanetData = (planetData) => {
  const planets = planetData.results.map( async (planet) => {
    const residents = await getPlanetResidents(planet.residents)

    return {
      name: planet.name,
      climate: planet.climate,
      terrain: planet.terrain,
      population: planet.population,
      residents: residents,
    }
  })

  return Promise.all(planets)
}