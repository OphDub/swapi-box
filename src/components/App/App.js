import React, { Component } from 'react';
import './App.css';
import Nav from '../Nav/Nav';
import Main from '../Main/Main';
import apiGet from '../api/api';

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      films: [],
      planets: [],
      people: [],
      vehicles: [],
      favorites: [],
    }
  }

  getFilmData = async () => {
    const films = await apiGet('films')

    this.setState({ films })
  }

  getData = async (request) => {
    const data = await apiGet(request)

    this.setState({ [request]: data })
  }

  componentDidMount() {
    this.getFilmData()
  }

  render() {
    return (
      <div className="App">
        <header>
          <h1>SWAPI BOX</h1>
        </header>
        <Nav  getData={this.getData}/>
        <Main film={this.state.films[0]}/>
      </div>
    );
  }
}

export default App;
