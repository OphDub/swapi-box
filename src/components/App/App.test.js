/* eslint-disable */
import React from 'react';
import { shallow, mount } from 'enzyme';
import App from './App';
import { apiGet } from '../helper/helper';
import {
  mockFilmData,
  mockCleanFilmData,
  mockPeopleData,
  mockCleanPeopleData,
  mockPlanetData,
  mockCleanPlanetData,
  mockVehicleData,
  mockCleanVehicleData
} from '../mock-data';

describe('APP', () => {
  it('should match the snapshot', async () => {
    window.fetch = jest.fn().mockImplementation(() => ({
      status: 200,
      json: () => new Promise((resolve, reject) => {
        resolve(mockFilmData)
      })
    }));

    const wrapper = await shallow(<App />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should have a default state with an empty film object and empty arrays for planets, people, vehicles, and favorites', async () => {
    window.fetch = jest.fn().mockImplementation(() => ({
      status: 200,
      json: () => new Promise((resolve, reject) => {
        resolve(mockFilmData)
      })
    }));

    const mockFn = jest.fn()
    const emptyObj = {};
    const emptyArray = [];
    const wrapper = await shallow(<App getRandomFilmData={mockFn}/>);

    expect(wrapper.state().film).toEqual(emptyObj);
    expect(wrapper.state().planets).toEqual(emptyArray);
    expect(wrapper.state().people).toEqual(emptyArray);
    expect(wrapper.state().vehicles).toEqual(emptyArray);
    expect(wrapper.state().favorites).toEqual(emptyArray);
  });

  it('should choose a random object from an array upon mounting', () => {

  });

  it('should be able to getData', () => {

  });

  it('should save objects to the favorites array and not add duplicates', async () => {
    window.fetch = jest.fn().mockImplementation(() => ({
      status: 200,
      json: () => new Promise((resolve, reject) => {
        resolve(mockFilmData)
      })
    }));

    const wrapper = await shallow(<App />);
    const result = [...mockCleanVehicleData, ...mockCleanPlanetData]

    wrapper.instance().saveFavorite(...mockCleanPlanetData)
    wrapper.instance().saveFavorite(...mockCleanVehicleData)
    wrapper.instance().saveFavorite(...mockCleanPlanetData)

    expect(wrapper.state().favorites).toEqual(result)
  });

  it('should not save the save object to the favorites array', () => {

  })

  it('should remove objects from the favorites array', () => {

  });
});
