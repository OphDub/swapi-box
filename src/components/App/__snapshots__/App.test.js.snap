// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`APP should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <header>
          <h1
                    className="swapi-title"
          >
                    SWAPI BOX
          </h1>
          <NavLink
                    activeClassName="active"
                    ariaCurrent="true"
                    to="/favorites"
          >
                    <button
                              className="faves"
                    >
                              Favorites
                              <div
                                        className="fave-count"
                              />
                    </button>
          </NavLink>
</header>,
        <Nav
          getData={[Function]}
/>,
        <Main
          favorites={Array []}
          film={Object {}}
          people={Array []}
          planets={Array []}
          saveFavorite={[Function]}
          vehicles={Array []}
/>,
      ],
      "className": "app",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h1
              className="swapi-title"
>
              SWAPI BOX
</h1>,
            <NavLink
              activeClassName="active"
              ariaCurrent="true"
              to="/favorites"
>
              <button
                            className="faves"
              >
                            Favorites
                            <div
                                          className="fave-count"
                            />
              </button>
</NavLink>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "SWAPI BOX",
              "className": "swapi-title",
            },
            "ref": null,
            "rendered": "SWAPI BOX",
            "type": "h1",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "activeClassName": "active",
              "ariaCurrent": "true",
              "children": <button
                className="faves"
>
                Favorites
                <div
                                className="fave-count"
                />
</button>,
              "to": "/favorites",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Favorites",
                  <div
                    className="fave-count"
/>,
                ],
                "className": "faves",
              },
              "ref": null,
              "rendered": Array [
                "Favorites",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": 0,
                    "className": "fave-count",
                  },
                  "ref": null,
                  "rendered": 0,
                  "type": "div",
                },
              ],
              "type": "button",
            },
            "type": [Function],
          },
        ],
        "type": "header",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "getData": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "favorites": Array [],
          "film": Object {},
          "people": Array [],
          "planets": Array [],
          "saveFavorite": [Function],
          "vehicles": Array [],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <header>
            <h1
                        className="swapi-title"
            >
                        SWAPI BOX
            </h1>
            <NavLink
                        activeClassName="active"
                        ariaCurrent="true"
                        to="/favorites"
            >
                        <button
                                    className="faves"
                        >
                                    Favorites
                                    <div
                                                className="fave-count"
                                    />
                        </button>
            </NavLink>
</header>,
          <Nav
            getData={[Function]}
/>,
          <Main
            favorites={Array []}
            film={Object {}}
            people={Array []}
            planets={Array []}
            saveFavorite={[Function]}
            vehicles={Array []}
/>,
        ],
        "className": "app",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h1
                className="swapi-title"
>
                SWAPI BOX
</h1>,
              <NavLink
                activeClassName="active"
                ariaCurrent="true"
                to="/favorites"
>
                <button
                                className="faves"
                >
                                Favorites
                                <div
                                                className="fave-count"
                                />
                </button>
</NavLink>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "SWAPI BOX",
                "className": "swapi-title",
              },
              "ref": null,
              "rendered": "SWAPI BOX",
              "type": "h1",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "activeClassName": "active",
                "ariaCurrent": "true",
                "children": <button
                  className="faves"
>
                  Favorites
                  <div
                                    className="fave-count"
                  />
</button>,
                "to": "/favorites",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "Favorites",
                    <div
                      className="fave-count"
/>,
                  ],
                  "className": "faves",
                },
                "ref": null,
                "rendered": Array [
                  "Favorites",
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": 0,
                      "className": "fave-count",
                    },
                    "ref": null,
                    "rendered": 0,
                    "type": "div",
                  },
                ],
                "type": "button",
              },
              "type": [Function],
            },
          ],
          "type": "header",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "getData": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "favorites": Array [],
            "film": Object {},
            "people": Array [],
            "planets": Array [],
            "saveFavorite": [Function],
            "vehicles": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
